pkg/machines/libvirt-dbus.es6

```diff
@@ -0,0 +1,861 @@
+                        name,
+                        connectionName,
+                        message: _("VM FORCE OFF action failed"),
+                        detail: { exception }
+                    })));
+        };
+    },
+
+    FORCEREBOOT_VM({
+        name,
+        connectionName,
+        id: objPath
+    }) {
+        return dispatch => {
+            clientLibvirt[connectionName].call(objPath, 'org.libvirt.Domain', 'Reset', [0], TIMEOUT)
+                    .fail(exception => dispatch(vmActionFailed({
+                        name,
+                        connectionName,
+                        message: _("VM FORCE REBOOT action failed"),
+                        detail: { exception }
+                    })));
+        };
+    },
+
+    /*
+     * Initiate read of all VMs
+     *
+     * @returns {Function}
+     */
+    GET_ALL_VMS({
+        connectionName,
+        libvirtServiceName
+    }) {
+        if (connectionName) {
+            return dispatch => {
+                dispatch(checkLibvirtStatus(libvirtServiceName));
+                clientLibvirt[connectionName] = cockpit.dbus('org.libvirt', { bus: connectionName });
+                startEventMonitor(dispatch, connectionName, libvirtServiceName);
+                doGetAllVms(dispatch, connectionName);
+                dispatch(getStoragePools(connectionName));
+                dispatch(getHypervisorMaxVCPU(connectionName));
+            };
+        }
+
+        return unknownConnectionName(getAllVms, libvirtServiceName);
+    },
+
+    GET_HYPERVISOR_MAX_VCPU({ connectionName }) {
+        logDebug(`${this.name}.GET_HYPERVISOR_MAX_VCPU: connection: ${connectionName}`);
+
+        if (connectionName) {
+            return dispatch => clientLibvirt[connectionName].call('/org/libvirt/QEMU', 'org.libvirt.Connect', 'GetDomainCapabilities', ['', '', '', '', 0], TIMEOUT)
+                    .done((capsXML) => {
+                        let count = getDomainMaxVCPU(capsXML[0]);
+                        dispatch(setHypervisorMaxVCPU({ count, connectionName }));
+                    })
+                    .fail(ex => console.error("GetDomainCapabilities failed: %s", ex));
+        }
+
+        return unknownConnectionName(setHypervisorMaxVCPU);
+    },
+
+    /**
+     * Retrieves list of libvirt "storage pools" for particular connection.
+     */
+    GET_STORAGE_POOLS({
+        connectionName
+    }) {
+        /*
+         * 2  -> VIR_CONNECT_LIST_STORAGE_POOLS_ACTIVE
+         * 64 -> VIR_CONNECT_LIST_STORAGE_POOLS_DIR
+         */
+        let flags = 2 | 64;
+        return dispatch => clientLibvirt[connectionName].call('/org/libvirt/QEMU', 'org.libvirt.Connect', 'ListStoragePools', [flags], TIMEOUT)
+                .done(objPaths => {
+                    let pools = [];
+
+                    logDebug(`GET_STORAGE_POOLS: object paths: ${JSON.stringify(objPaths)}`);
+
+                    function get_storage_pool(i) {
```

Comment:
The clean way to avoid that in the code is to wait with the `.setState()` call until all parallel promises are finished (using `Promise.all`). If that is somehow unwieldy, the test should have some appropriate `wait_present()` statements (e. g. wait for the n-th row in the table - the test should be able to predict how many rows it expects)

Reply:
So, testing can be trickier sometimes;
I have a test that it's sometimes failing and basically does the following;

* click Add Disk to existing VM
* Click Use Existing Disk
* Choose Storage Pool
* Click the dropdown button to choose storage Volume.
Errors on storage volumes list being empty for specific pool.

I can never reproduce manually.
The issue that probably I hit here is that the test is getting to the button to click the dropdown for volumes before the state is updated so that it contains the storage volumes list.
The dropdown button will never be updated by react while it's still clicked.
So ```wait_present()``` will not help.